### ISSUE TITLE
Training is not happening with train.py file

### ISSUE THREAD

**Issue by sachinsharma9780** [fabvio/ld-lsi#9]
![epocsh](https://user-images.githubusercontent.com/40523048/93472647-40705100-f8f5-11ea-8cb9-ceeee22ab902.png)
 I am using the train.py file to just train the network on BDD dataset with 10 epochs but the results are almost same for each epoch. I made some changes in train.py file to train it for 10 epochs. Training is performed on just 10% of data. Could you please have look ?

@khizar-anjum 
@fabvio 

Q2) Where in the code we are passing labels for road classification?

Here is the code:

import torch
import torchvision
import os
from glob import glob
from PIL import Image
from utils import AverageMeter, iou, debug_val_example
from tqdm import tqdm
import numpy as np

class CrossEntropyLoss2d(torch.nn.Module):
    #NLLLoss2d is negative log-likelihood loss
    #it returns the semantic segmentation cross-entropy loss2d
    def __init__(self, weight=None):
        super().__init__()
        self.loss = torch.nn.NLLLoss(weight)

    def forward(self, outputs, targets):
        outputs, _ = outputs
        return self.loss(torch.nn.functional.log_softmax(outputs, dim=1), targets) 

def train(model, criterion, optimizer, data_loader, debug_data_loader, device, ntrain_batches):
	model.train()
	top1 = AverageMeter('Acc@1', ':6.2f')
	avgloss = AverageMeter('Loss', '1.5f')

	cnt = 0
	for data in tqdm(data_loader):
	    cnt += 1
	    model.to(device)
	    #print('.',end='')
	    image, target = data['image'].to(device), data['label'].squeeze(1)
	    target = torch.round(target*255) #converting to range 0-255
	    target = target.type(torch.int64).to(device)
	    output = model(image)
	    loss = criterion(output, target)
        #  clears old gradients from the last step (otherwise you’d just accumulate the gradients from all loss.backward() calls).
	    optimizer.zero_grad()
	    # computes the derivative of the loss w.r.t. the parameters (or anything requiring gradients) using backpropagation
	    loss.backward()
	    # performs a parameter update based on the current gradient and update rule (for ex sgd)
	    optimizer.step()
	    acc1 = iou(output, target)
	    #print("iou", acc1)
	    top1.update(acc1, image.size(0))
	    avgloss.update(loss, image.size(0))
	    # if cnt%10 == 0:
	    #     #debug val example checks and prints a random debug example every
	    #     #10 batches.
	    #     #it uses cpu, so might be slow. 
	    #     #debug_val_example(model, debug_data_loader)
	    #     print('Loss', avgloss.avg)

	    #     print('Training: * Acc@1 {top1.avg:.3f}'
	    #                 .format(top1=top1))
	        
	    # if cnt >= ntrain_batches:
	    #     return 
	print('Loss', avgloss.avg)
	print('Training: * Acc@1 {top1.avg:.3f}'
	                    .format(top1=top1))
	# print('Full imagenet train set:  * Acc@1 {top1.global_avg:.3f}'
	#             .format(top1=top1))
	return

class berkely_driving_dataset(torch.utils.data.Dataset):
    def __init__(self, path, type='train', transform=None, color = True):
    # dataloader for bdd100k segmentation dataset
    # path should contain the address to bdd100k folder
    # it generally has the following diretory structure
        """
        - bdd100k
          - drivable_maps
            - color_labels
            - labels
          - images
            - 100k
            - 10k
        """
        # type can either be 'train' or 'val'
        self.path = path
        self.type = type
        self.transform = transform
        self.imgs = glob(os.path.join(self.path, 'images/100k/' + self.type + '/*.jpg'))
        if color is True:
            print(' True color')
            self.labels = [os.path.join(self.path, 'drivable_maps/color_labels/' + self.type,\
                        x.split('/')[-1][:-4] + '_drivable_color.png') for x in self.imgs]
        else:
            print('Color False')
            self.labels = [os.path.join(self.path, 'drivable_maps/labels/' + self.type,\
                                     x.split('/')[-1][:-4]+'_drivable_id.png') for x in self.imgs]
            #print('labels', self.labels)
        self.length = len(self.imgs)

    # so that len(dataset) returns the size of the dataset.
    def __len__(self):
        return self.length

   #  to support the indexing such that dataset[i] can be used to get ith sample 
    def __getitem__(self, idx):
        if torch.is_tensor(idx):
            idx = idx.tolist()

        #print('idx', idx)
        img_name = self.imgs[idx]
        image = Image.open(img_name)
        label = Image.open(self.labels[idx])
        if self.transform:
            image = self.transform(image)
            label = self.transform(label)
        #print('img', image, 'label', label.shape)
        return {'image':image, 'label':label}

if __name__=='__main__':
    """
    Here is an example of how to use these functions for training
    This script is designed to train on berkely driving dataset. Therefore, the 
    PATH_TO_BERKELY_DATASET variable points to the root of that dataset. You might
    have to edit it.
    """

    #DEFINING SOME IMPORTANT VARIABLES
    PATH_TO_BERKELY_DATASET = '/home/sachin/Desktop/sampled_data/bdd100k'

    #loading libraries
    from models import erfnet_road

    #loading models
    roadnet = erfnet_road.Net()
    if torch.cuda.is_available():
        device = torch.device("cuda")

    else:
        device = torch.device("cpu")
      

    loading weights
    model_w = torch.load('/home/sachin/Desktop/ld-lsi/res/weights/weights_erfnet_road.pth')
    new_mw = {}
    for k,w in model_w.items():
        new_mw[k[7:]] = w
    roadnet.load_state_dict(new_mw)

    roadnet.to(device)
    roadnet.eval();

    # Making Dataloaders
    bdd_transforms = torchvision.transforms.Compose([
        torchvision.transforms.Resize((360, 640)),
        torchvision.transforms.RandomHorizontalFlip(),
        torchvision.transforms.ToTensor()
    ])

    bdd_train = berkely_driving_dataset(PATH_TO_BERKELY_DATASET, transform=bdd_transforms,  type='train', color = False)
    bdd_val = berkely_driving_dataset(PATH_TO_BERKELY_DATASET, transform=bdd_transforms,  type='val', color = False)

    sampler_train = torch.utils.data.RandomSampler(bdd_train)
    print('train shape', sampler_train)
    sampler_val = torch.utils.data.RandomSampler(bdd_val)
    print('val shape', sampler_val)

    dl_train = torch.utils.data.DataLoader(
        bdd_train, batch_size=32,
        sampler=sampler_train)
    
    #for data in dl_train:
     #   print('k', data['image'].shape) 
    # the valiation only works with a batchsize of 1
    dl_val = torch.utils.data.DataLoader(
        bdd_val, batch_size=32,
        sampler=sampler_val)

    #defining losses
    criterion = CrossEntropyLoss2d()
    optimizer = torch.optim.Adam(roadnet.parameters(), 10e-3)

    #training an epoch for 100 batches
    for epoch in range(0, 11):
    	print("epoch:", epoch)
    	train(roadnet, criterion, optimizer, dl_train, dl_val, device, 100)

**Comment by khizar-anjum** [comment]
@sachinsharma9780, as far as I can see, the loss value is decreasing, which means that the network is being trained. Instead of posting all the code (which is overwhelming and distracting), can you just post the areas which you changed? That way, I can look at them and figure out what is happening.

**Comment by sachinsharma9780** [comment]
Thank You very much @khizar-anjum  for replying. Here are the changes which I made:

1.) In the train method:

![train](https://user-images.githubusercontent.com/40523048/93616589-345bc080-f9d5-11ea-85d9-f703d10b1eec.png)


sampler_train = torch.utils.data.RandomSampler(bdd_train)

dl_train = torch.utils.data.DataLoader(
        bdd_train, batch_size=32,
        sampler=sampler_train)

for epoch in range(0, 20):
    	print("epoch:", epoch)
    	train(roadnet, criterion, optimizer, dl_train, dl_val, device, 100)

Q2) Why we are not normalizing images?

NOTE: I am just trying to train the network and not validating it. Also, I am training on 10% data, although on full data also pattern remains the same.

**Comment by sachinsharma9780** [comment]
My second ques is, where we are passing labels for road classification forex residential, highway etc in the code?

**Comment by sachinsharma9780** [comment]
> @sachinsharma9780, as far as I can see, the loss value is decreasing, which means that the network is being trained. Instead of posting all the code (which is overwhelming and distracting), can you just post the areas which you changed? That way, I can look at them and figure out what is happening.

**Comment by fabvio** [comment]
> My second ques is, where we are passing labels for road classification forex residential, highway etc in the code?

I'll answer to this. Basically, if you look at the paper where the code is from, the task was joint free space detection in multiple lanes and road type recognition, because it was needed for a simple planning algorithm which had to take into account different lane change policies. You can change lane if you are in highway, while doing that in a residential row requires more attention.

**Comment by sachinsharma9780** [comment]
> > My second ques is, where we are passing labels for road classification forex residential, highway etc in the code?
> 
> I'll answer to this. Basically, if you look at the paper where the code is from, the task was joint free space detection in multiple lanes and road type recognition, because it was needed for a simple planning algorithm which had to take into account different lane change policies. You can change lane if you are in highway, while doing that in a residential row requires more attention.

@fabvio : Yeah I read your paper thoroughly and it's talk about Multi-Task learning i.e. free space detection and road type classification which is very interesting approach. But what I am asking is to train this kind of network we need to provide labels to the network that this input frame belongs to let's say residential-class but in the given "training" code, I cannot see where the author is passing these labels to the network.

Thank You!

**Comment by khizar-anjum** [comment]
Hi, so sachinsharma9780, you have to understand that the training code is unofficial and I posted it myself while reading the paper. It was just meant to help other people who were trying to write the same. 

I did not pass the labels to the network in the training code. So, you would have to add that functionality yourself. Please open a PR if you add it.

**Comment by sachinsharma9780** [comment]
@khizar-anjum  Hi Thanks for replying. Have you seen the training script
screenshot is it fine? What I am not getting is why the training acc is not
increasing although loss is decreasing. Can it be the case that semantic
labels which we are passing are not one hot encoded just have contiguous
ids in the shape 360*640.

Another thing which I am thinking is instead of passing black images what
if we pass color ground truths by setting color=True. Since I have seen in
some works they do the same. Will that make difference?

**Comment by khizar-anjum** [comment]
Yes, please try to pass colored ground truths as well. I think there may be some bug in calculating accuracy, because it should change if the loss is decreasing.

**Comment by sachinsharma9780** [comment]
@fabvio , @khizar-anjum : 

I have found one bug in your training code due to which it was giving  wrong accuracy.

![bug](https://user-images.githubusercontent.com/40523048/94049313-1cb97900-fdd5-11ea-8214-e901f30bf80a.png)

While calculating iou you are adding one additional dimension to the output via unsqueeze operation which is making issues while calculating iou. Therefore after removing it training acc is increasing.

Note: Above screenshot is for erfnet and not for erfnet_road. For erfnet_road you have to do:
 output, x = output
output = torch.argmax(output, 1)

**Comment by khizar-anjum** [comment]
I see. Thanks for pointing it out. Can you post a PR for correcting the bug?

**Comment by sachinsharma9780** [comment]
ok. I will do that.

**Comment by sachinsharma9780** [comment]
@khizar-anjum @fabvio 

I have found one more bug in the training code while calculating iou score. 
get_score function in utils.py file calculate "logical and" operation between target and output which I think is not the correct approach because lets say we have 2*2 matrix for prediction and target containing values :

target = [[0, 1,  1, 2]] and pred = [[0, 1, 2, 1]]

In this case, background class accuracy will never be included which is fine in most of the cases. But, when we will perform a 
![bug2](https://user-images.githubusercontent.com/40523048/94181864-5f945300-fea0-11ea-93e1-5213cd9c41e4.png)
"logical and " operation between 1 and 2 the result will be "TRUE" which implies ego lanes are correctly classified as other lanes or vice versa, the same situation for logical and between 2 and 1. Therefore, it is adding the wrong results to the intersection variable. 

Hence, I have used Jaccard index metrics from sklearn. Below you can see the difference between these 2 metrics when I train my code:
![bug2_score](https://user-images.githubusercontent.com/40523048/94182087-b437ce00-fea0-11ea-9629-f0f6f3bea3b7.png)

Let me know, If I am wrong because this is very important.

I would also encourage @fabvio to give his valuable feedback.

Note: Results are on 10% data.

**Comment by hamidreza-abrar** [comment]
hi 
i have problem with import rospy in cnn node. 
i have installed rospkg but still have problem.
it says no module named rospy . 
how can i solve this???

**Comment by sachinsharma9780** [comment]
@fabvio 
I have fixed all the bugs and can achieve the IOU mentioned in the paper. In addition to this, I have also added my own experiments. I did this work as part of my Master Thesis to improve your findings.

Here is the link to my GitHub repository: https://github.com/sachinsharma9780/Master-Thesis-Free-Space-Detetction-for-Autonomous-Driving-using-Deep-Learning.

**Comment by khizar-anjum** [comment]
This is great @sachinsharma9780. Good luck for your Masters thesis!
